[FAB-18120] Add integration test for DevMode using new lifecycle

DereckLuo opened the pull request:
#### Type of change

- Test update

#### Description
* Added integration test support for DevMode using new lifecycle.
* Added DevMode support for ChaincodeEndorsementInfoSource.

#### Related issues
[FAB-18120](https://jira.hyperledger.org/browse/FAB-18120)
[FAB-18169](https://jira.hyperledger.org/browse/FAB-18169)

ghaskins commented:
Is this covered by the unit-tests?  If so, is there a way I can run it locally to verify since CI doesn't currently trigger sdk tests?

adecaro commented:
Hi @ghaskins, please, find in the README the instructions on how to test TLS connection in dev mode.  Please, let me know if they are clear enough.

binhn commented:
@ghaskins currently unit test requires manual set up per readme until deploy is implemented.

sykesm reviewed (changes requested):
I think most of the comments on the 1.4 version of these changes apply here as well. PTAL at those comments, apply as appropriate, and I'll take another peek.

Thanks.

DereckLuo committed:
[FAB-18169] Add DevMode support in ChaincodeEndorsementInfoSource

* Added DevMode support in ChaincodeEndorsementInfoSource, when
  chaincode is launched by developer and missing Installinfo from
  chaincodeInfo cache, rebuild Installinfo and return accordinly.

Signed-off-by: Chongxin Luo <Chongxin.Luo@ibm.com>

DereckLuo committed:
[FAB-18120] Adding DevMode integration test for new lifecycle.

* Added integration test for DevMode using new lifecycle.

Signed-off-by: Chongxin Luo <Chongxin.Luo@ibm.com>

wlahti commented on integration/nwo/deploy.go:
```diff
@@ -242,7 +241,11 @@ func ApproveChaincodeForMyOrg(n *Network, channel string, orderer *Orderer, chai
 				ClientAuth:          n.ClientAuthRequired,
-			})
+			}
+			if !p.DevMode {
```
I think we should invert this to match the typical usage (i.e. all networks outside of the devmode suite don't use devmode): always set the packageID above but set it to "" when devmode is enabled. What do you think?

DereckLuo replied:
good point, done.

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,291 @@
+		By("enable new lifecycle capabilities")
+		nwo.EnableCapabilities(network, "testchannel", "Application", "V2_0", orderer, org1peer0)
+
+		By("packaging chaincode")
```
We don't need to package the chaincode anymore since we aren't installing anything in this updated version of the test, right?

DereckLuo replied:
it turns out `SetPackageIDFromPackageFile` in `ApproveChaincodeForMyOrg` is looking for packaged chaincode file to obtain chaincode PackagedID. 
changed the logic slightly, and fixed it.

wlahti commented on integration/nwo/deploy.go:
```diff
@@ -242,7 +241,11 @@ func ApproveChaincodeForMyOrg(n *Network, channel string, orderer *Orderer, chai
+			}
+			if p.DevMode {
+				cmd.PackageID = ""
+			}
```
We don't need this after your update, right? And then we can revert the session and command code above to its original state.

DereckLuo replied:
yes, correct.

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,277 @@
+		os.RemoveAll(testDir)
+	})
+
+	It("executing chaincode in dev mode using legacy lifecycle", func() {
```
```suggestion
	It("executes chaincode in dev mode using legacy lifecycle", func() {
```
Super nit

DereckLuo replied:
fixed

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,277 @@
+		Expect(docContainers[0].Names[0]).To(ContainSubstring("Org1.peer0-mycc-0.0"))
+	})
+
+	It("executing chaincode in dev mode", func() {
```
```suggestion
	It("executes chaincode in dev mode", func() {
```
Super nit round 2

DereckLuo replied:
fixed

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,277 @@
+		By("setting up the channel")
+		network.CreateAndJoinChannel(orderer, channelName)
+
+		By("enable new lifecycle capabilities")
```
```suggestion
		By("enabling V2_0 application capabilities")
```
I'd prefer for this to match what's actually being set on the channel, which isn't lifecycle-specific. Still a nit though. :)

DereckLuo replied:
fixed

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,277 @@
+		chaincodeProcess = ifrit.Invoke(chaincodeRunner)
+		Eventually(chaincodeProcess.Ready(), network.EventuallyTimeout).Should(BeClosed())
+
+		By("approving chaincode for orgs")
```
```suggestion
		By("approving the chaincode definition")
```
Nit. Also feels like we should remove "orgs" since we're using a single org network now.

DereckLuo replied:
fixed

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,277 @@
+
+		By("approving chaincode for orgs")
+		nwo.ApproveChaincodeForMyOrg(network, channelName, orderer, chaincode, peers...)
+		By("committing chaincode")
```
```suggestion
		By("committing the chaincode definition")
```
Nit :)

DereckLuo replied:
fixed

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,277 @@
+		chaincodeProcess.Signal(syscall.SIGKILL)
+		Eventually(chaincodeProcess.Wait(), network.EventuallyTimeout).Should(Receive())
+
+		By("invoking chaincode after it is being killed")
```
```suggestion
		By("invoking chaincode after it has been killed, expecting it to fail")
```

DereckLuo replied:
fixed

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,277 @@
+	Eventually(sess, n.EventuallyTimeout).Should(gexec.Exit(0))
+	Expect(sess).To(gbytes.Say(strconv.Itoa(queryValue)))
+
+	By("invoke the chaincode")
```
```suggestion
	By("invoking the chaincode")
```

DereckLuo replied:
fixed

wlahti reviewed (commented):
One change (in approveformyorg) and a number of nits, but otherwise this is looking good from my perspective.

wlahti commented on integration/nwo/deploy.go:
```diff
@@ -220,13 +220,12 @@ func InstallChaincodeLegacy(n *Network, chaincode Chaincode, peers ...*Peer) {
 	// used to ensure we only approve once per org
 	approvedOrgs := map[string]bool{}
 	for _, p := range peers {
+		if chaincode.PackageID == "" && !p.DevMode {
```
Taking the bigger picture into account, I think we shouldn't have this check for DevMode in this nwo helper function for approving the chaincode definition. I'd only expect devmode to be utilized in the devmode suite so I think it makes the most sense to simply include a similar helper function there and keep the nwo package's helper function streamlined.

DereckLuo replied:
fixed.

wlahti commented:
@sykesm Can you take another look when you get a chance? Things have changed fairly significantly since your last review. Thanks!

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		// turn on dev mode for all peers
+		for _, p := range network.Peers {
+			p.DevMode = true
+		}
```
There should be only one peer. Why not just change it?

`network.Peer("Org1", "peer0").DevMode = true`

DereckLuo replied:
fixed.

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		soloConfig := nwo.BasicSolo()
+		soloConfig.RemoveOrg("Org2")
```
Please don't add to the config structure. They are meant to be literals. (There's one other method on config and it really needs to be removed.)

Instead, you can create a network configuration for this package that does exactly what you want and nothing else. This one works for me:

```go
var devNetwork = &nwo.Config{
	Organizations: []*nwo.Organization{{
		Name:   "OrdererOrg",
		MSPID:  "OrdererMSP",
		Domain: "example.com",
		CA:     &nwo.CA{Hostname: "ca"},
	}, {
		Name:          "Org1",
		MSPID:         "Org1MSP",
		Domain:        "org1.example.com",
		EnableNodeOUs: true,
		Users:         1,
		CA:            &nwo.CA{Hostname: "ca"},
	}},
	Consortiums: []*nwo.Consortium{
		{Name: "DevConsortium", Organizations: []string{"Org1"}},
	},
	Consensus: &nwo.Consensus{
		Type:            "solo",
		BootstrapMethod: "file",
	},
	SystemChannel: &nwo.SystemChannel{
		Name:    "systemchannel",
		Profile: "OrderingProfile",
	},
	Channels: []*nwo.Channel{
		{Name: "devchannel", Profile: "DevProfile"},
	},
	Profiles: []*nwo.Profile{
		{Name: "OrderingProfile", Orderers: []string{"orderer"}},
		{Name: "DevProfile", Consortium: "DevConsortium", Organizations: []string{"Org1"}},
	},
	Orderers: []*nwo.Orderer{
		{Name: "orderer", Organization: "OrdererOrg"},
	},
	Peers: []*nwo.Peer{{
		Name:         "peer0",
		Organization: "Org1",
		Channels:     []*nwo.PeerChannel{{Name: "devchannel"}},
	}},
}
```

DereckLuo replied:
fixed.

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		peerChaincodeAddress := network.PeerAddress(org1peer0, nwo.ChaincodePort)
+		envs := []string{
+			"CORE_PEER_TLS_ENABLED=false",
+			fmt.Sprintf("CORE_CHAINCODE_ID_NAME=%s", legacyChaincodeID),
```
Appending is a good choice here: `"CORE_CHAINCODE_ID_NAME=" + legacyChaincodeID`

DereckLuo replied:
fixed

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		chaincodeProcess = ifrit.Invoke(chaincodeRunner)
+		Eventually(chaincodeProcess.Ready(), network.EventuallyTimeout).Should(BeClosed())
+
+		docContainers, err := client.ListContainers(docker.ListContainersOptions{All: false})
```
nit: 'doc' implies documentation, not docker.

non-nit: Why are we playing with Docker at all? If this is dev mode, the chaincode should be started and registered by the developer. If it goes down, the chaincode should not be started by the peer - which appears to be what is being tested.

At the end of the day, the behavior from the new chaincode is what I'd expect here: the peer *does not* start (non-system) chaincode when we're in dev mode. That means all of the docker related enumerations and validation can be removed.

wlahti replied:
We added this because we couldn't think of another way to truly prove that we're using our devmode chaincode process and not just a docker container started by the peer. With that said, this exact check is at a point that doesn't make sense. If anything, it should be after the instantiation call, since i'm pretty sure the init baked into instantiate is when the container is started with legacy chaincodes. 

In general though, we'll defer to your judgment on whether to include any docker checks or not.

sykesm replied:
Issue a message when the chaincode starts and on every request received when a flag (not normally provided by the peer) or an environment variable (not normally provided by the peer) is set. Then make assertions on the output of the chaincode.

No docker. Low complexity. Easy to implement and addresses the readiness check comments.

wlahti replied:
Ah right, similar to what we did for the _lifecycle test to prove we actually stopped chaincodes when they were no longer referenced. Sounds good. We'll handle this particular comment, @DereckLuo.

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		cmd := exec.Command(chaincodeExecutePath, []string{"-peer.address", peerChaincodeAddress}...)
+		cmd.Env = append(cmd.Env, envs...)
+		chaincodeRunner = ginkgomon.New(ginkgomon.Config{
+			Name:              "chaincode",
+			Command:           cmd,
+			StartCheckTimeout: 15 * time.Second,
```
If there's no start check so the timeout doesn't matter and the `Eventually(chaincodeProcess.Ready())` assertions are useless. When the start check is not specified, it becomes immediately satisfied.

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+
+		docContainers, err := client.ListContainers(docker.ListContainersOptions{All: false})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(docContainers).Should(HaveLen(0))
```
Expect/To Eventually/Should (here and elsewhere)

DereckLuo replied:
fixed. and else where

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+
+		docContainers, err = client.ListContainers(docker.ListContainersOptions{All: false})
+		Expect(docContainers).Should(HaveLen(1))
+		Expect(docContainers[0].Names[0]).To(ContainSubstring("Org1.peer0-mycc-0.0"))
```
Please see above: I believe this is asserting incorrect behavior that we do not want to preserve in v2

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		network.CreateAndJoinChannel(orderer, channelName)
+
+		By("enabling V2_0 application capabilities")
+		nwo.EnableCapabilities(network, "testchannel", "Application", "V2_0", orderer, org1peer0)
```
You're using `channelName` everywhere else. I had to change this when I used the hand-crafted network.

DereckLuo replied:
fixed

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		legacyChaincodeID := fmt.Sprintf("%s:%s", legacyChaincode.Name, legacyChaincode.Version)
+
+		org1peer0 := network.Peer("Org1", "peer0")
+		peers := network.PeersWithChannel(channelName)
```
This confused me since we have exactly one peer. Again, with one peer, I'd expect to use the explicit peer instance that we're testing.

Later on this is used in the variadic list for no effect as we already explicitly use org1peer0.

DereckLuo replied:
fixed

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+			"CORE_PEER_TLS_ENABLED=false",
+			fmt.Sprintf("CORE_CHAINCODE_ID_NAME=%s", legacyChaincodeID),
+		}
+		cmd := exec.Command(chaincodeExecutePath, []string{"-peer.address", peerChaincodeAddress}...)
```
Why the complicated slice and expansion instead of just providing the two arguments?

DereckLuo replied:
fixed

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		}
+
+		org1peer0 := network.Peer("Org1", "peer0")
+		peers := network.PeersWithChannel(channelName)
```
Same as the other test: We have one peer so I don't understand why we need this list and the expansions later on.

DereckLuo replied:
fixed

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		envs := []string{
+			"CORE_PEER_TLS_ENABLED=false",
+			fmt.Sprintf("CORE_PEER_ADDRESS=%s", peerChaincodeAddress),
+			fmt.Sprintf("CORE_CHAINCODE_ID_NAME=%s", chaincodeID),
```
No need to use `Sprintf` and when we're just concatenating strings.

DereckLuo replied:
fixed

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		nwo.EnableCapabilities(network, "testchannel", "Application", "V2_0", orderer, org1peer0)
+
+		By("running the chaincode")
+		chaincodeID := fmt.Sprintf("%s:%s", chaincode.Name, chaincode.Version)
```
No need to use `Sprintf` when we're concatenating three strings.

`chaincode.Name + ":" + chaincodeVersion`

DereckLuo replied:
fixed

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+			Ctor:    `{"Args":["init","a","100","b","200"]}`,
+			Policy:  `OR ('Org1MSP.member')`,
+		}
+		legacyChaincodeID := fmt.Sprintf("%s:%s", legacyChaincode.Name, legacyChaincode.Version)
```
No need to use `Sprintf` to concatenate three strings.

`legacyChaincode.Name + ":" + legacyChaincode.Version` ?

Also, it's way up here but not used until the environment is built.

DereckLuo replied:
fixed

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+			fmt.Sprintf("CORE_CHAINCODE_ID_NAME=%s", chaincodeID),
+		}
+
+		cmd := exec.Command(chaincode.Path, []string{"-peer.address", peerChaincodeAddress}...)
```
Why the complicated slice + expansion instead of just providing the two arguments?

DereckLuo replied:
fixed

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,305 @@
+		chaincodeRunner = ginkgomon.New(ginkgomon.Config{
+			Name:              "chaincode",
+			Command:           cmd,
+			StartCheckTimeout: 15 * time.Second,
```
Same comment about start check. Timeout is irrelevant if you don't have a check.

sykesm commented on integration/nwo/config.go:
```diff
@@ -29,3 +29,41 @@ func (c *Config) RemovePeer(orgName, peerName string) {
+		}
+		pro.Organizations = orgs
+	}
+}
```
Please remove this in favor of defining a fit-for-purpose network.

DereckLuo replied:
fixed.

sykesm reviewed (changes requested):
Lots of nits. The biggest actions are

- Please remove the logic added to the network literal configuration structures. The configurations are meant to be built up for tests. My recommended configuration was added to a comment.
- Please do not assert on the broken behavior of the legacy lifecycle starting chaincode when we're in "dev mode". We should probably discuss fixing that but I certainly don't want to preserve that behavior as it's unexpected and inconsistent. (Imagine working on the code, restarting the process, and then finding it's no longer the chaincode under development that you're running...)
- There's lots of code here that assumes multiple peers even though we explicitly have one.

sykesm commented on integration/nwo/standard_networks.go:
```diff
@@ -219,3 +219,59 @@ func MultiNodeEtcdRaft() *Config {
+		},
+		SystemChannel: &SystemChannel{
+			Name:    "systemchannel",
+			Profile: "TwoOrgsOrdererGenesis",
```
No longer TwoOrgs but one.

wlahti replied:
Fixed

sykesm commented on integration/nwo/standard_networks.go:
```diff
@@ -219,3 +219,59 @@ func MultiNodeEtcdRaft() *Config {
+			{Name: "orderer", Organization: "OrdererOrg"},
+		},
+		Channels: []*Channel{
+			{Name: "testchannel", Profile: "TwoOrgsChannel"},
```
No longer TwoOrgs but one.

wlahti replied:
Fixed

sykesm commented on integration/nwo/standard_networks.go:
```diff
@@ -219,3 +219,59 @@ func MultiNodeEtcdRaft() *Config {
+			},
+		}},
+		Profiles: []*Profile{{
+			Name:     "TwoOrgsOrdererGenesis",
```
No longer TwoOrgs but one.

wlahti replied:
Fixed

sykesm commented on integration/nwo/standard_networks.go:
```diff
@@ -219,3 +219,59 @@ func MultiNodeEtcdRaft() *Config {
+			Name:     "TwoOrgsOrdererGenesis",
+			Orderers: []string{"orderer"},
+		}, {
+			Name:          "TwoOrgsChannel",
```
No longer TwoOrgs but one.

wlahti replied:
Fixed

sykesm reviewed (commented):
Other than org names (implying two instead of one development org), the pending comments are related to the docker assertions and start checks. Other than that, looks good to me.

Thanks.

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,287 @@
+		By("querying and invoking the chaincode")
+		RunQueryInvokeQuery(network, orderer, org1peer0, channelName, 90)
+
+		Eventually(chaincodeRunner.Buffer()).Should(gbytes.Say("invoking in devmode"))
```
Shouldn't this check happen above after the first queryinvokequery?

stephyee replied:
🧹

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,287 @@
+		By("querying and invoking the chaincode")
+		RunQueryInvokeQuery(network, orderer, org1peer0, channelName, 100)
+
+		By("killing chaincode process")
```
I believe this is the part Matt was saying we shouldn't assert on since it's unusual (broken?) behavior to have the legacy lifecycle still start up the chaincode for you in devmode here. Let's remove lines 125-130?

stephyee replied:
🧹  Removed

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,287 @@
+		}
+
+		By("querying and invoking the chaincode")
+		RunQueryInvokeQuery(network, orderer, org1peer0, channelName, 100)
```
Where's the Buffer() check here?

stephyee replied:
🧹  Added

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,287 @@
+		chaincodeRunner = ginkgomon.New(ginkgomon.Config{
+			Name:              "chaincode",
+			Command:           cmd,
+			StartCheckTimeout: 15 * time.Second,
```
StartCheck (and related chaincode update) coming soon?

stephyee replied:
🧹

wlahti commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,286 @@
+			StartCheckTimeout: 15 * time.Second,
+			StartCheck:        "starting up in devmode...",
+		})
+		chaincodeProcess = ifrit.Invoke(chaincodeRunner)
```
Are we missing "Eventually(chaincodeProcess.Ready(), network.EventuallyTimeout).Should(BeClosed())" here?

wlahti replied:
I'll handle this and one nit that Matt mentioned a while ago (moving the network configuration out of the standard_networks).

wlahti replied:
Fixed

sykesm commented on core/chaincode/lifecycle/endorsment_info.go:
```diff
@@ -90,6 +92,12 @@ func (cei *ChaincodeEndorsementInfoSource) CachedChaincodeInfo(channelID, chainc
 	if chaincodeInfo.InstallInfo == nil {
+		if cei.UserRunsCC {
+			chaincodeInfo.InstallInfo = &ChaincodeInstallInfo{
+				PackageID: fmt.Sprintf("%s:%s", chaincodeName, chaincodeInfo.Definition.EndorsementInfo.Version),
```
In a follow-on, please consider replacing the printf with concatenation.

wlahti replied:
Done

sykesm commented on integration/chaincode/simple/cmd/main.go:
```diff
@@ -48,6 +48,10 @@ func handleSigTerm() {
+	if os.Getenv("DEVMODE_ENABLED") != "" {
+		fmt.Println("starting up in devmode...")
+	}
+
```
If this is being used as the readiness check, shouldn't it happen after starting the go routine that calls `shim.Start`?  Neither is perfect but it seems closer to what we'd expect.

wlahti replied:
Done

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,340 @@
+		network = nwo.New(devModeSolo, testDir, client, StartPort(), components)
+
+		network.TLSEnabled = false
+		// turn on dev mode for all peers
```
comment seems stale.

wlahti replied:
Done

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,340 @@
+
+		By("querying and invoking the chaincode")
+		RunQueryInvokeQuery(network, orderer, org1peer0, channelName, 100)
+		Eventually(chaincodeRunner.Buffer()).Should(gbytes.Say("invoking in devmode"))
```
a `ginkgomon.Runner` implements `gbytes.BufferProvider` so the call to `Buffer` is not needed.

wlahti replied:
Done

sykesm commented on integration/devmode/devmode_test.go:
```diff
@@ -0,0 +1,340 @@
+
+		By("querying and invoking the chaincode")
+		RunQueryInvokeQuery(network, orderer, org1peer0, channelName, 100)
+		Eventually(chaincodeRunner.Buffer()).Should(gbytes.Say("invoking in devmode"))
```
Same comment about BufferProvider.

wlahti replied:
Done

sykesm reviewed (dismissed):
I'll let you decide if you want to make any other changes here or in a separate PR.

wlahti commented:
@Mergifyio backport release-2.2

lindluni commented:
@Mergifyio backport release-2.2
